lib/internal/crypto/pbkdf2.js

```diff
@@ -27,51 +43,95 @@ function pbkdf2(password, salt, iterations, keylen, digest, callback) {
   if (typeof callback !== 'function')
     throw new ERR_INVALID_CALLBACK(callback);
 
-  const encoding = getDefaultEncoding();
-  const keybuf = Buffer.alloc(keylen);
+  const job = new PBKDF2Job(
+    kCryptoJobAsync,
+    password,
+    salt,
+    iterations,
+    keylen,
+    digest);
 
-  const wrap = new AsyncWrap(Providers.PBKDF2REQUEST);
-  wrap.ondone = (ok) => {  // Retains keybuf while request is in flight.
-    if (!ok) return callback.call(wrap, new ERR_CRYPTO_PBKDF2_ERROR());
-    if (encoding === 'buffer') return callback.call(wrap, null, keybuf);
-    callback.call(wrap, null, keybuf.toString(encoding));
+  const encoding = getDefaultEncoding();
+  job.ondone = (err, result) => {
+    if (err !== undefined)
+      return callback.call(job, err);
+    const buf = Buffer.from(result);
+    if (encoding === 'buffer')
+      return callback.call(job, null, buf);
+    callback.call(job, null, buf.toString(encoding));
```

Comment:
```suggestion
      return FunctionPrototypeCall(callback, job, err);
    const buf = Buffer.from(result);
    if (encoding === 'buffer')
      return FunctionPrototypeCall(callback, job, null, buf);
    FunctionPrototypeCall(callback, job, null, buf.toString(encoding));
```
not sure if BufferFrom or BufferPrototypeToString are things, but if so, those would be needed here too

Reply:
They are not given that they are not primordials.

Reply:
It's still a node primordial, and it'd still be good for node to be robust against `delete Buffer.prototype.toString`, for example, but if there's not already a pattern for it, ¯\\\_(ツ)_/¯

Reply:
It's something that can be tackled in a separate PR. As it stands now, there are about 64 `Buffer.from()` uses in lib, 38 `Buffer.alloc()` or `Buffer.allocUnsafe()`, about 57 `.toString()` calls that aren't necessarily Buffer but could be... If we are going to treat Buffer as a primitive (we I agree we should) then we should do that in a separate PR.